### PR TITLE
Add dict hash search instead of linear search of static headers.

### PR DESCRIPTION
**Pull request by KhasanovBI**
Add dict hash search instead of linear search of static headers.

### CODE REVIEW

File: hpack/table.py [hunk 1]
```diff
@@ -172,12 +185,15 @@ def search(self, name, value):
...
-                elif partial is None:
-                    partial = (i + 1, n, None)
+
+        header_name_search_result = HeaderTable.STATIC_TABLE_MAPPING.get(name)
+        if header_name_search_result is not None:
+            index = header_name_search_result.positions.get(value)
+            if index is not None:
+                return index + 1, name, value
+            else:
+                partial = (header_name_search_result.min + 1, name, None)
```
Comment (KhasanovBI) [inline hpack/table.py]: dict search: 
```python
In [26]: %timeit ht.search(b':authority', b'')
The slowest run took 12.41 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 557 ns per loop

In [27]: %timeit ht.search(b'www-authenticate', b'')
The slowest run took 13.68 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 593 ns per loop
```
linear search:
```python
In [31]: %timeit ht.search(b':authority', b'')
The slowest run took 11.18 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 618 ns per loop

In [32]: %timeit ht.search(b'www-authenticate', b'')
The slowest run took 9.54 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 4.2 µs per loop
```

File: hpack/table.py [hunk 2]
```diff
@@ -23,6 +23,12 @@ def table_entry_size(name, value):
     return 32 + len(name) + len(value)
 
 
+class _HeaderNameSearchResult(object):
+    def __init__(self):
+        self.min = None
+        self.positions = {}
```
Comment (KhasanovBI) [inline hpack/table.py]: Could you explain more, I did not quite understand what was meant. PS. Sorry for my english.

Comment (Lukasa) [inline hpack/table.py]: Sorry, this was mostly just a question for myself and the other maintainers: it shouldn't affect your pull request. :grin:

File: hpack/table.py [hunk 3]
```diff
@@ -105,6 +111,13 @@ class HeaderTable(object):
 
     STATIC_TABLE_LENGTH = len(STATIC_TABLE)
 
+    STATIC_TABLE_MAPPING = {}
+    for index, (name, value) in enumerate(STATIC_TABLE):
+        header_name_search_result = STATIC_TABLE_MAPPING.setdefault(name, _HeaderNameSearchResult())
+        header_name_search_result.positions[value] = index
+        min_index = header_name_search_result.min
+        header_name_search_result.min = min(min_index, index) if min_index is not None else index
```
Comment (KhasanovBI) [inline hpack/table.py]: Move to staticmethod, IMHO more convenient way is use metaclass at this case, but for python2/3 compatibility metaclass require some magic like `six.with_metaclass`. I can rewrite if the current choise does not suit.

Comment (Lukasa) [inline hpack/table.py]: Don't use a `staticmethod`: there's no need. Let's just write an actual function: not all code needs to be attached to classes.

File: hpack/table.py (hunk 1 above)
Comment (SethMichaelLarson) [inline hpack/table.py]: I don't see any problems with this approach performance-wise. Might even be worth it to examine a trie-based approach and see if that could be faster.

Comment (Lukasa) [inline hpack/table.py]: I would be surprised if a pure-Python trie was faster than a C dict on any implementation other than PyPy.

Comment (Lukasa) [inline hpack/table.py]: Do we have an idea of what the memory overhead is here?

Comment (SethMichaelLarson) [inline hpack/table.py]: You can build a trie using dicts, I'm suggesting getting rid of the custom object that holds onto a 'min' value and positions and instead using `name` and `value` as keys within a two-nested dictionary. Is there any reason that can't be done?

Comment (Lukasa) [inline hpack/table.py]: Oh interesting, that's not what I'd considered doing, but yeah, that could work. The overhead of that gets a bit higher again, and I wonder about what we would do with dynamic entries. But it could probably be managed.

Comment (SethMichaelLarson) [inline hpack/table.py]: Dynamic entries would be more difficult to simulate "pushing" down the table but probably could be done. 2 hash lookups plus whatever is necessary to track updates compared to iterating over a header table, unpacking tuples, and checking values. Might be something to think about?

I was just suggesting using this method for the static table where you don't have to track changes.

Comment (Lukasa) [inline hpack/table.py]: I feel like that ends up with a pretty large overhead of dicts...

Comment (SethMichaelLarson) [inline hpack/table.py]: Would it not be the same amount of dicts that's being used in this PR minus the `_HeaderNameSearchResult` objects? There's one dict with names and a dict per name with values contained in the `_HeaderNameSearchResult` instance.

Perhaps we could minimize to one dict where you attempt `d[(name, value)]` first and if that fails then you attempt `d[(name, None)]`?

Comment (Lukasa) [inline hpack/table.py]: Hrm.

I think what'd be best is if we could benchmark all those side-by-side. Get a feel for what the relative costs of each implementation are. Memory consumption of them would also be good. :grin:

**Comment by SethMichaelLarson** [comment]
Here are my findings based on these static header table implementations (lower is always better):
Tested on Python 3.5.2 w/ 5,000,000 lookups for each test

Size of the static table: 7,496 bytes

### Tuple Dict Method

Structure size: 9,024 bytes
Successful Name, Value Lookup: 1.717s
Partial Name, Value Lookup: 2.569s
Missed Name, Value Lookup: 2.386s

### Proposed Method

Structure size: 31,312 bytes
Successful Name, Value Lookup: 2.007s
Partial Name, Value Lookup: 2.198s
Missed Name, Value Lookup: 1.402s

From this data I'd conclude that if memory problems aren't our concern then I would recommend the proposed method in this pull request due to the much lower partial/miss lookup times.

File: hpack/table.py (hunk 1 above)
Comment (SethMichaelLarson) [inline hpack/table.py]: Did some testing and I included my analysis below.

**Comment by Lukasa** [comment]
Hrm. I don't want to penalise successful lookups too much: they'll happen a lot. If our memory overhead is static (that is, if we don't allocate a new collection of data structures for each encoder/decoder) then I'm happy to accept the overhead.

**Comment by SethMichaelLarson** [comment]
I'm not sure of the proportion of successful / partial / miss lookups on the header table. If I knew that information I could give a much better analysis of these benchmarks and how it'll affect us.

On a positive note any of these methods are much faster than what we're currently doing. :tada:

**Comment by Lukasa** [comment]
We can work that out too: run the tests that use the test fixtures in this repo! They're a reasonable snapshot of what other HPACK implementations do, and can also be used to test against ourselves and to test encoding.

**Comment by SethMichaelLarson** [comment]
I'll reimplement these into a branch of master and see how fast they can run the test suite.

**Comment by KhasanovBI** [comment]
Hi) So what we do with PR, can you summarize necessary changes please?

**Comment by Lukasa** [comment]
Right now we're kinda hoping to work out what the performance

**Comment by Lukasa** [comment]
Right now we're hoping to work out what version of this code performs best on the corpus of test fixtures in the module.

**Comment by SethMichaelLarson** [comment]
Yeah sorry kind of dropped the ball on this. Too many other things!

**Comment by SethMichaelLarson** [comment]
## Results

### Current Method: 59.04 seconds

### Modified Method: 55.60 seconds

```python
    def search(self, name, value):
        """
        Searches the table for the entry specified by name
        and value

        Returns one of the following:
            - ``None``, no match at all
            - ``(index, name, None)`` for partial matches on name only.
            - ``(index, name, value)`` for perfect matches.
        """
        result = STATIC_TABLE_MAPPING.get(name)
        partial = None
        if result:
            min_index, pos = result
            index = pos.get(value)
            if index is not None:
                return index, name, value
            else:
                partial = (min_index, name, None)

        offset = STATIC_TABLE_LENGTH + 1

        for i, (n, v) in enumerate(self.dynamic_entries):
            if n == name:
                if v == value:
                    return i + offset, n, v
                elif partial is None:
                    partial = (i + offset, n, None)

        return partial
```

I had to change a few tests that directly used `STATIC_TABLE` and moved all the `STATIC_*` to globals within `table.py` in order to achieve this.

**Comment by Lukasa** [comment]
Ok cool @SethMichaelLarson, are you recommending some changes to this PR then?

**Comment by SethMichaelLarson** [comment]
It'd be best to move away from objects into tuples to hold onto the data within the `STATIC_TABLE_MAPPING` and move the `offset` calculation using `STATIC_TABLE_LENGTH` until after the static table is checked. I also moved all the `index + 1` calculations into the function constructing the `STATIC_TABLE_MAPPING`.

File: hpack/table.py [hunk 4]
```diff
@@ -172,12 +172,15 @@ def search(self, name, value):
         """
         offset = HeaderTable.STATIC_TABLE_LENGTH + 1
```
Comment (SethMichaelLarson) [inline hpack/table.py]: Move this calculation to after the `STATIC_TABLE_MAPPING` is checked.

File: hpack/table.py [hunk 5]
```diff
@@ -213,3 +216,43 @@ def _shrink(self):
             cursize -= table_entry_size(name, value)
             log.debug("Evicting %s: %s from the header table", name, value)
         self._current_size = cursize
+
+
+class _HeaderNameSearchResult(object):
```
Comment (SethMichaelLarson) [inline hpack/table.py]: Instead of using an `object`, use a `tuple` with the same attributes.

File: hpack/table.py [hunk 6]
```diff
@@ -213,3 +216,43 @@ def _shrink(self):
...
+
+    static_table_mapping used for hash searching.
+    """
+    static_table_mapping = {}
+    for index, (name, value) in enumerate(HeaderTable.STATIC_TABLE):
+        header_name_search_result = static_table_mapping.get(name)
+        if not header_name_search_result:
+            header_name_search_result = _HeaderNameSearchResult()
+            static_table_mapping[name] = header_name_search_result
+        header_name_search_result.positions[value] = index
```
Comment (SethMichaelLarson) [inline hpack/table.py]: Store `index + 1` rather than just `index` as it's always used as `index + 1` within `HeaderTable.search()`

File: hpack/table.py [hunk 7]
```diff
@@ -213,3 +216,43 @@ def _shrink(self):
...
+
+        min_index = header_name_search_result.min
+        if min_index is not None:
+            new_min_index = min(min_index, index)
```
Comment (SethMichaelLarson) [inline hpack/table.py]: Same as above here.

File: hpack/table.py [hunk 8]
```diff
@@ -213,3 +216,43 @@ def _shrink(self):
...
+        else:
+            new_min_index = index
+        header_name_search_result.min = new_min_index
```
Comment (SethMichaelLarson) [inline hpack/table.py]: Same as above here.

**Comment by SethMichaelLarson** [comment]
@Lukasa I left my review comments for changes. Thanks @KhasanovBI for the patience and for tackling this! :tada:

File: hpack/table.py [hunk 9]
```diff
@@ -170,14 +170,17 @@ def search(self, name, value):
...
         partial = None
-        for (i, (n, v)) in enumerate(HeaderTable.STATIC_TABLE):
-            if n == name:
-                if v == value:
-                    return i + 1, n, v
-                elif partial is None:
-                    partial = (i + 1, n, None)
+
+        header_name_search_result = HeaderTable.STATIC_TABLE_MAPPING.get(name)
+        if header_name_search_result is not None:
```
Comment (SethMichaelLarson) [inline hpack/table.py]: `if x is not None` is slower than `if x` as long as we know that the two options for this `dict.get()` are either a `tuple` or `None` so we can make this optimization.

Comment (sigmavirus24) [inline hpack/table.py]: @SethMichaelLarson how much slower? In reality, I doubt it matters that much compared to other portions of this.

Comment (SethMichaelLarson) [inline hpack/table.py]: From my past experimentation on the order of nano seconds so yes much slower.

Comment (sigmavirus24) [inline hpack/table.py]: If we're concerned with *nanoseconds* then hpack must be in an incredibly fast place performance-wise. (For context, 1 nanosecond is 0.000000001 second)

Comment (sigmavirus24) [inline hpack/table.py]: Ah you edited your comment making me look like a jerk. 😄

Comment (SethMichaelLarson) [inline hpack/table.py]: Sorry! I was hoping you hadn't read it because what I wrote was the complete opposite intent of what I meant to say. I really need to audit my comments before I snap-post. :stuck_out_tongue:

**Comment by SethMichaelLarson** [comment]
You should also add yourself to `CONTRIBUTORS.rst` @KhasanovBI :)

**Comment by SethMichaelLarson** [comment]
Congrats! Thanks for this PR @KhasanovBI! :tada:

**Comment by KhasanovBI** [comment]
Hooray) Thanks for good review:) Perhaps we should look at the implementation for cython in the future for more speedup.
